packages/contracts-bedrock/src/L1/OPContractsManager.sol

```diff
@@ -401,6 +426,84 @@ contract OPContractsManager is ISemver {
+            });
+
+            if (IOptimismPortal2(payable(opChainAddrs.optimismPortal)).superchainConfig() != superchainConfig) {
+                revert SuperchainConfigMismatch();
+            }
+
+            IProxyAdmin proxyAdmin = _opChains[i].proxyAdmin;
+
+            // -------- Upgrade Contracts Stored in SystemConfig --------
+
+            // forgefmt: disable-start
+            proxyAdmin.upgrade(payable(address(systemConfig)), impls.systemConfigImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.l1CrossDomainMessenger), impls.l1CrossDomainMessengerImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.l1ERC721Bridge), impls.l1ERC721BridgeImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.l1StandardBridge), impls.l1StandardBridgeImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.disputeGameFactory), impls.disputeGameFactoryImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.optimismPortal), impls.optimismPortalImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.optimismMintableERC20Factory), impls.optimismMintableERC20FactoryImpl);
+
+            // -------- Discover and Upgrade Proofs Contracts --------
+            IPermissionedDisputeGame permissionedDisputeGame = IPermissionedDisputeGame(
+                address(getGameImplementation(IDisputeGameFactory(opChainAddrs.disputeGameFactory), GameTypes.PERMISSIONED_CANNON))
+            );
+            IFaultDisputeGame faultDisputeGame = IFaultDisputeGame(
+                address(getGameImplementation(IDisputeGameFactory(opChainAddrs.disputeGameFactory), GameTypes.CANNON))
+            );
+            IDelayedWETH delayedWETHPermissionlessGameProxy;
+            if (address(faultDisputeGame) != address(0)) {
+                delayedWETHPermissionlessGameProxy = faultDisputeGame.weth();
+            }
+
+            OpChainDiscoverables memory discoverables =  OpChainDiscoverables({
+                l2ChainId: permissionedDisputeGame.l2ChainId(),
+                anchorStateRegistryProxy: permissionedDisputeGame.anchorStateRegistry(),
+                faultDisputeGame: faultDisputeGame,
+                permissionedDisputeGame: permissionedDisputeGame,
+                delayedWETHPermissionedGameProxy: permissionedDisputeGame.weth(),
+                delayedWETHPermissionlessGameProxy: delayedWETHPermissionlessGameProxy
+            });
+
+
+            proxyAdmin.upgrade(payable(address(discoverables.anchorStateRegistryProxy)), impls.anchorStateRegistryImpl);
+            proxyAdmin.upgrade(payable(address(discoverables.delayedWETHPermissionedGameProxy)), impls.delayedWETHImpl);
+
+            // Not all chains have the permissionlessDelayedWETH
+            if (address(discoverables.delayedWETHPermissionlessGameProxy) != address(0)) {
+                proxyAdmin.upgrade(payable(address(discoverables.delayedWETHPermissionlessGameProxy)), impls.delayedWETHImpl);
+            }
```

Comment:
Clean up: 

- remove discoverables struct
- combine if(fdg) branches together, they are equivalent.
- create internal function for upgrade calls

Reply:
this is all done.